### PR TITLE
Interactive span selector improvement

### PR DESCRIPTION
**Pull request by ericpre**
## PR Summary

This PR updates the `SpanSelector` to make it consistent with `RectangleSelector`:
- Deprecate `span_stays` in favour of `interactive`
- Add edge handle when `interactive=True`
- Add `drag_from_anywhere` option, in line with #19657
- Fix https://github.com/matplotlib/matplotlib/pull/9660
- Improve docstring
- Deprecate `pressv`
- When reviewing the docstrings, I realised that some attributes are not expected to be used by users and should be private - I am happy to revert it back if necessary
  - Privatize `rect`, `rectprops`,  for the `SpanSelector`
  - Privatize  `to_draw`, `drawtype` for the `RectangleSelector
  - Privatize `active_handle` for SpanSelector and RectangleSelector
- Make `direction` a property

If you are happy with the changes, I will do the remaining doc, examples, etc updates.

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/widgets.py
Comment (anntzer) [inline lib/matplotlib/widgets.py:2131]: should this be private?

File: lib/matplotlib/widgets.py
Comment (anntzer) [inline lib/matplotlib/widgets.py:2089]: This should be fixed (perhaps in the base class implementation) so that one can set extents *before* the widget is ever drawn (e.g. to start with an already drawn SpanSelector).  Currently, setting `extents` before the first `draw` causes `update` to raise `AttributeError: draw_artist can only be used after an initial draw which caches the renderer` (it's probably just a matter of checking whether _cachedRenderer is set and skip the draw in that case.

File: lib/matplotlib/widgets.py
Comment (anntzer) [inline lib/matplotlib/widgets.py:2005]: This should use `@_api.rename_parameter` to generate a standard deprecation message and to error out if e.g. the user passes both `span_stays` *and* `interactive`.

**Comment by anntzer** [comment]
`interactive` differs from `span_stays` in that with `span_stays` it was possible to completely unset a previously set SpanSelector by clicking anywhere (effectively drawing a zero-sized selector), whereas this is now impossible(?) with `interactive`.  Not vetoing the change, but I would be slightly annoyed, being used to the old semantics (well, I don't know how intuitive they were either)...

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py:2131]: Yes, I agree that this should be private. I didn't make it private to be consistent with the API of `RectangleSelector` and `EllipseSelector`. Instead, would it be better to make it private for all 3 with a `@_api.deprecated('3.5')` in the case of `RectangleSelector` and `EllipseSelector`?

**Comment by ericpre** [comment]
> `interactive` differs from `span_stays` in that with `span_stays` it was possible to completely unset a previously set SpanSelector by clicking anywhere (effectively drawing a zero-sized selector), whereas this is now impossible(?) with `interactive`. Not vetoing the change, but I would be slightly annoyed, being used to the old semantics (well, I don't know how intuitive they were either)...
> (I guess the old semantics are conflicting with drag_from_anywhere :/ but at least they should beep working when `drag_from_anywhere = False` (the default))

Thanks for the review! Indeed, this is a change in behaviour and the motivation for this really is to have the `SpanSelector` behaving consistently with other widgets where possible. The current inconsistency could be considered as a bug! ;)
My view on this is that consistency and improved code readability/easier maintenance are worth a (small?) change in behaviour!

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py]: Done

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py:2080]: Done

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py]: Went ahead and privatise for all 3.

**Comment by ericpre** [comment]
> > `interactive` differs from `span_stays` in that with `span_stays` it was possible to completely unset a previously set SpanSelector by clicking anywhere (effectively drawing a zero-sized selector), whereas this is now impossible(?) with `interactive`. Not vetoing the change, but I would be slightly annoyed, being used to the old semantics (well, I don't know how intuitive they were either)...
> > (I guess the old semantics are conflicting with drag_from_anywhere :/ but at least they should beep working when `drag_from_anywhere = False` (the default))
> 
> Thanks for the review! Indeed, this is a change in behaviour and the motivation for this really is to have the `SpanSelector` behaving consistently with other widgets where possible. The current inconsistency could be considered as a bug! ;)
> My view on this is that consistency and improved code readability/easier maintenance are worth a (small?) change in behaviour!

Actually, thinking a bit more about it, I would like to add a separate optional parameter to the `SpanSelector` (and `RectangleSelector`, etc) to ignore events outside the widgets and I would prefer to do this in a follow up PR, because there are already a lot of changes in this PR. What do you think?

File: lib/matplotlib/widgets.py
Comment (ianhi) [inline lib/matplotlib/widgets.py:1981]: I think you're changing the order of the arguments here which is backwards incompatible. Someone may be (inadvisably) using positional args for every param and this will break that.

Can you switch the order so that all new kwargs are added at the end.

Comment (ericpre) [reply lib/matplotlib/widgets.py]: Done, I have left `interactive` at the same position as `span_stays` since it replace it.

**Comment by ericpre** [comment]
Rebase on master to check that #9660 is fixed with this PR and #20264 and it seems to be the case!

**Comment by jklymak** [comment]
@anntzer @ianhi, did this get where you wanted?

**Comment by anntzer** [comment]
AFAICT, with the new design, it is not possible to completely get rid of an (interactive/span_stays) SpanSelector?  Previously this could be done by clicking anywhere, effectively drawing a SpanSelector smaller than minspan which destroyed the old one.

**Comment by ericpre** [comment]
> AFAICT, with the new design, it is not possible to completely get rid of an (interactive/span_stays) SpanSelector? Previously this could be done by clicking anywhere, effectively drawing a SpanSelector smaller than minspan which destroyed the old one.

@anntzer: indeed, I have tried to answer to this comment [above](https://github.com/matplotlib/matplotlib/pull/20113#issuecomment-840665611). Also, I would add that this behaviour is not documented! ;) The documented way of removing the `SpanSelector` is to set `span_selector.active` to False (https://matplotlib.org/stable/api/widgets_api.html#matplotlib.widgets.SpanSelector).

As mentioned previously, I am convinced that the inconsistency between the different selector is confusing for users and would justify a change in behaviour. Also, this PR make the implementation of `SpanSelector` very similar to the `RectangleSelector` one, which will make the maintenance of selectors easier. Considering that their maintenance is particularly not straightforward (difficult to test fully, backend dependent, blitting dependent, etc), I think this is worth it.

What do you think?

**Comment by anntzer** [comment]
To be fair, I don't think I can make a really objective judgment here because I am used to the old behavior and actually like it better.  But if you convince enough other core devs that the new behavior is better, I won't block the change either.

File: lib/matplotlib/widgets.py
Comment (ianhi) [inline lib/matplotlib/widgets.py:2249]: Can you add a `Parameters` section describing what the valid inputs are.

File: lib/matplotlib/widgets.py
Comment (ianhi) [inline lib/matplotlib/widgets.py:2260]: `Parameters` and `Returns` section again. Should specify things like what coordinate system `x` and `y` are expected to be in. Also, what do you mean by the `index`? The index of the handles?

File: lib/matplotlib/widgets.py
Comment (ianhi) [inline lib/matplotlib/widgets.py:2257]:
```suggestion
def set_animated(self, val):
    for m in self._markers:
        m.set_animated(val)
```

otherwise you get `AttributeError: 'list' object has no attribute 'set_animated'`

File: lib/matplotlib/widgets.py
Comment (ianhi) [inline lib/matplotlib/widgets.py:2254]:
```suggestion
    def set_visible(self, val):
        for m in self._markers:
             m.set_visible(val)
```

same attribute error as `set_animated`

File: lib/matplotlib/widgets.py
Comment (ianhi) [inline lib/matplotlib/widgets.py:2197]: Can we use a name other than `vmin` and `vmax`? Those are very clearly associated with colormaps and i think I would be confused to see them here.

**Review by ianhi — CHANGES_REQUESTED** [review]
I'm :+1:  on this change. I strongly prefer the new behavior (I admittedly wasn't using the old one all that often) and like the consistency with RectangleSelector.


Changes I would like to see:

- More meaningful naming with `markers` vs `lines` (see comments)
- More robust docstrings (see comments) especially for things such as `set_data`
- Thoughts on making `drag_from_anywhere` default true if `interactive=True`? (see comment)
- An inclusion of these in an example
     - Given the consistency between Rectangle and Span maybe there can be a single large `*Selector` example that goes over attributes they share?
     - I'm less clear on if this is a good idea than the former points

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py]: Done

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py]: Done

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py]: Done

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py:1961]: I agree that this is the most natural behaviour, however, I don't understand how it would not change the current default (`drag_from_anywhere` doesn't exist and clicking inside the selector does nothing), your suggestion means that the current default will change when `interactive` is True and now it will move the selector.  With`interactive=False`, `drag_from_anywhere` doesn't do anything and I don't see how having `drag_from_anywhere=None` can help here!

**Review by ericpre — COMMENTED** [review]
@ianhi: all comments addressed.

**Comment by ericpre** [comment]
> To be fair, I don't think I can make a really objective judgment here because I am used to the old behavior and actually like it better. But if you convince enough other core devs that the new behavior is better, I won't block the change either.

@anntzer, as mentioned [above](https://github.com/matplotlib/matplotlib/pull/20113#issuecomment-840665611), I am happy to do this in a follow up PR, where I would like to add the option for a selector to ignore events outside of itself. The idea would be to be able to create several selector in the same axes.

File: lib/matplotlib/widgets.py
Comment (ianhi) [reply lib/matplotlib/widgets.py:1961]: Sorry what I meant is to have a line like this:
```python
drag_from_anywhere = interactive if drag_from_anywhere is None else drag_from_anywhere
```

> d not change the current default (drag_from_anywhere doesn't exist and clicking inside the selector does nothing),

I suppose this is a behavior change, but I was considering it more as a UI change than an API change and thus  perhaps escaping the woes of backwards incompat - though I'm not the expert on where that line falls.

**Comment by ericpre** [comment]
I have updated the [PR summary](https://github.com/matplotlib/matplotlib/pull/20113#issue-626341847) for the latest changes.

@dstansby, since you have been working on the `RectangleSelector` recently, it would be good if you could comment on the privatisation of some attributes of the `RectangleSelector` proposed in this PR, thanks!

**Comment by jklymak** [comment]
ping @ianhi and @anntzer for reviews on this....

**Comment by jklymak** [comment]
Discussed on the call.  @anntzer is still not in favour for the reasons discussed above.  @tacaswell will look at soon and render some sort of decision on how to proceed.   Thanks @ericpre for your patience on this!

File: lib/matplotlib/widgets.py
Comment (tacaswell) [reply lib/matplotlib/widgets.py:1961]: The line depends on how angry it is likely to make a user who was expecting the old behavior and got the new one! 

I'm with @anntzer that I have used the old behavior (you can not move, only re-select and clicking gives a "zero width" selection which removes the selector), but still optimistic that we can find a middle ground.

Comment (tacaswell) [reply lib/matplotlib/widgets.py:1961]: I took the liberty of pushing a commit to add back the ability to dismiss the selection by clicking somewhere else in the plot.

**Comment by tacaswell** [comment]
I am a big fan of the new functionality!  Being able to grab the edge and "tweak" the range or the whole slider and drag it are both super nice.

But like @anntzer I think that losing the ability to effectively dismiss the selection is too high a cost so I figured out how to put it back (see f10ab5c , I hope that you do not mind I did this @ericpre !  I had the code to make this work and pushing the changes seemed better than pasting a diff here or trying to explain how I did it.  Please force-push out of existence if you do not like the changes).

We need API change notes for all of the deprecation/privitizations.

**Review by tacaswell — COMMENTED** [review]
Provisional approval, contingent on some version of restoring the "clearing" being implemented and API change notes.

Between this and @ianhi 's review I'll say this has 1 approval so this can be merged by anyone.

**Comment by ericpre** [comment]
Thanks @tacaswell for the changes, I updated it slightly to avoid drawing the span selector.
I have updated the doc and now it should be good to go!

File: lib/matplotlib/widgets.py
Comment (QuLogic) [inline lib/matplotlib/widgets.py:2297]: How can you be sure that 0 in the other coordinate doesn't produce some invalid transform (e.g., if the scale is log)?

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py:2108]: Yes, it was intentional and one reason for that is that it was fully working before (span wasn't drawn at all). In f64e98b, I change it to allow changing direction of an already existing span selector. It makes the code slightly more complicated and I can revert it if we think that this is not worth it.

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py]: I privatised `interactive`. The alternative would be to make it a property, would it be better?

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py]: Normally, it is one or another, depending if the span selector is vertical or horizontal, so we need to loop over only one of the two.

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py:2345]: I don't know but I couldn't find an issue. Do you have an example?

On the topic of log scale, it seems that `drag_from_anywhere` doesn't behave as we would expect and the same issue occurs with `RectangleSelector`/`EllipseSelector`.

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py:1961]: > ```python
> drag_from_anywhere = interactive if drag_from_anywhere is None else drag_from_anywhere
> ```

This is equivalent to have `drag_from_anywhere=True` by default because when `interactive=False`, `drag_from_anywhere` is not used and therefore this is a behaviour change: if the user uses `span_stays=True` which is now `interactive=True` and they expect not to move the span selector when clicking on it.

**Comment by ericpre** [comment]
@QuLogic, thanks for the review, I have addressed all points!

File: lib/matplotlib/widgets.py
Comment (QuLogic) [reply lib/matplotlib/widgets.py]: I meant, can one of those functions replace the loop? It might not be the case, since the return type is different.

Comment (ericpre) [reply lib/matplotlib/widgets.py]: OK, you meant something along these lines:
```python
import matplotlib.pyplot as plt

fig = plt.figure()
ax = fig.add_subplot()
ax.plot(range(100))

positions = (10, 20)

# instead of having a loop over the positions, pass the positions tuple as argument
ax.axvline(x=positions)
```

However, the first argument of `ax.axvline` or `ax.axhline` expect a float and we can't pass an iterable. The alternative would be to use `ax.vlines` or `ax.hlines` with a transform for the infinite direction, but it will more complicated.

**Comment by ericpre** [comment]
@anntzer, @ianhi, @tacaswell, @QuLogic, is there anything left to do on this PR?

File: lib/matplotlib/widgets.py
Comment (QuLogic) [inline lib/matplotlib/widgets.py:1952]:
```suggestion
        when *interactive* is True. See `matplotlib.lines.Line2D` for details
```

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py:2040]: Actually, it is not clear to me: `pressv` and `prev` are deprecated and not only made private, in this cases which is better to use, `_api.deprecated` or `_api.deprecate_privatize_attribute`?

Comment (QuLogic) [reply lib/matplotlib/widgets.py:2040]: `_api.deprecate_privatize_attribute` is simply a wrapper around `_api.deprecated` with an automatically-derived property implementation. There is no semantic difference other than shorter code.

File: lib/matplotlib/widgets.py
Comment (QuLogic) [inline lib/matplotlib/widgets.py:1993]: Did you mean to use the property setter here? Otherwise, `direction` is not checked on `__init__`.

File: lib/matplotlib/widgets.py
Comment (QuLogic) [reply lib/matplotlib/widgets.py:2346]: I tried it out in the example, and it seems to be fine for log scale in the other coordinate. It must just produce some silent NaNs.

Log scale on the span selector coordinate works "as expected" in that it's moved in the data space, and gets bigger/smaller based on the position in log space. I don't know if we want to fix it or not (or if it's really a fix exactly.)

**Comment by anntzer** [comment]
(Not that it adds much to the discussion as @tacaswell mostly summarized my points above, but I think that I finally realized why I am so attached to the old UI model: it mimicks the UI of text selection, where you can restart another selection by just clicking anywhere and selecting from there.)

File: lib/matplotlib/widgets.py
Comment (ericpre) [reply lib/matplotlib/widgets.py]: No, I didn't, I suspect that the property setter have been later to allow changing direction of an existing selector and this hasn't been updated.

**Comment by ericpre** [comment]
> (Not that it adds much to the discussion as @tacaswell mostly summarized my points above, but I think that I finally realized why I am so attached to the old UI model: it mimicks the UI of text selection, where you can restart another selection by just clicking anywhere and selecting from there.)

I see what you mean, however I would imagine that with the recent addition of `drag_from_anywhere`, changing the selection area is much easier and that it covers this case to a certain extent.

File: lib/matplotlib/widgets.py
Comment (anntzer) [reply lib/matplotlib/widgets.py:2348]: It feels to me as if we should move in screen space instead?

Comment (QuLogic) [reply lib/matplotlib/widgets.py:2348]: I don't think any widgets do that (e.g., see discussion on #19864).

Comment (anntzer) [reply lib/matplotlib/widgets.py:2348]: OK, I guess we can keep that discussion for another time then (and move them all at once).

Comment (ericpre) [reply lib/matplotlib/widgets.py:2348]: Yes, and it would need to be done for all selectors.

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
@anntzer I think the current state captures the best of the old behavior (click outside of the selection region and make a new span) and the new (you can drag either edge and if you click on the middle you drag the whole span).  Both of the new interaction modes are things that I personally wished for in my past-life (to either scan a fixed-width window across some line or to tweak it to be _just_ the right width).

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:1946]: To me the primary difference is whether the selector is still visible after the first selection.

When `interactive=True` I get a selector that persists, but I can still modify later. Coming from that context I would expect `interactive=False` to still show my selected region, but I cannot modify it anymore. This indicates that `interactive` is a misleading name.

When starting from scratch, I'd call this `persistent`, but given that we have `span_stays` keeping that is ok as well.

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:1951]: Since there's no canonical line in a span. the name does not make it clear what this controls. I'd call these `handle_props` or `handle_line_props` if you want to be very explicit.

File: lib/matplotlib/widgets.py
Comment (timhoffm) [inline lib/matplotlib/widgets.py:1952]:
```suggestion
        Properties of the handler lines at the edges of the span. Only used
```

**Review by timhoffm — CHANGES_REQUESTED** [review]

**Comment by timhoffm** [comment]
:roll_eyes: Too late. Will open a follow-up PR.

**Comment by ericpre** [comment]
Thanks everyone for the reviews, it has been very useful!

**Comment by timhoffm** [comment]
Thanks @ericpre for keeping up!

**Comment by larsoner** [comment]
In MNE we allow users to update the selector colors, and have:
```
        selector.rect.set_color(color)
        selector.rectprops.update(dict(facecolor=color))
```
This now emits a deprecation warning from this PR, but it doesn't indicate how code should be migrated. I looked at this and #20558 and hoped / thought maybe I could find some `selector.set_*` or `selector.handle_props` or something to modify but I don't see anything. How should we update our code? (And it might be worth improving the deprecation warning to give some hints to anyone else who hits this issue.) Maybe with something like this?
```
selector.artists[0].set_color(color)
```
the `artists[0]` is the `selector.rect`. But this just seems like a hack workaround, and if there are properties held internally it will not "stick" so I'm guessing it's not the right idea...

**Comment by jklymak** [comment]
@larsoner suggest you open an issue.

**Pull request merged**
